imap_processing/mag/l1c/interpolation_methods.py

```diff
@@ -94,10 +119,101 @@ def cubic(
     return spline(output_timestamps)
 
 
+def estimate_rate(timestamps: np.ndarray) -> VecSec:
+    """
+    Given a set of timestamps, estimate the rate of the timestamps.
+
+    This rate will be one of the defined rates in the VecSec enum. The calculation
+    assumes there are no significant gaps in the timestamps.
+
+    Parameters
+    ----------
+    timestamps : numpy.ndarray
+        1D array of timestamps to estimate the rate of.
+
+    Returns
+    -------
+    VecSec
+        Estimated rate of the timestamps.
+    """
+    samples_per_second = timestamps.shape[0] / (timestamps[-1] - timestamps[0]) * 1e9
+    per_second = VecSec(
+        POSSIBLE_RATES[(np.abs(POSSIBLE_RATES - samples_per_second)).argmin()]
+    )
+
+    return per_second
+
+
+def cic_filter(
+    input_vectors: np.ndarray,
+    input_timestamps: np.ndarray,
+    output_timestamps: np.ndarray,
+    input_rate: Optional[VecSec],
+    output_rate: Optional[VecSec],
+):
+    """
+    Apply CIC filter to data before interpolating.
+
+    The filtering uses a Cascaded integrator-comb (CIC) filter which is used in FSW to
+    filter down the raw data to telemetered data.
+
+    This assumes that the input_vectors and input_timestamps are downsampled to
+    the output_timestamps rate. Neither input_timestamps nor output_timestamps should
+    have significant gaps.
+
+    Parameters
+    ----------
+    input_vectors : numpy.ndarray
+        Input vectors of shape (n, 3) where n is equal to the number of input
+        timestamps. Contains x, y, z components of the vector.
+    input_timestamps : numpy.ndarray
+        Input timestamps of shape (n,) which correspond to the timestamps of the input
+        vectors.
+    output_timestamps : numpy.ndarray
+        Output timestamps of shape (m,) to generate interpolated vectors for.
+    input_rate : VecSec, optional
+        Expected rate of input timestamps.
+    output_rate : VecSec, optional
+        Expected rate of output timestamps.
+
+    Returns
+    -------
+    numpy.ndarray
+        Filtered something something.
+    """
+    # output rate should always be higher
+    input_rate = estimate_rate(input_timestamps) if input_rate is None else input_rate
+    output_rate = (
+        estimate_rate(output_timestamps) if output_rate is None else output_rate
+    )
+
+    if input_rate.value <= output_rate.value:
+        raise ValueError(
+            f"Burst mode input rate {input_rate} should never be less than "
+            f"the normal mode output rate {output_rate}. "
+            f"Both rates are required"
+        )
+
+    decimation_factor = int(input_rate.value / output_rate.value)
+    cic1 = np.ones(decimation_factor)
+    cic1 = cic1 / decimation_factor
+    cic2 = np.convolve(cic1, cic1)
+    delay = (len(cic2) - 1) // 2
+    input_filtered = input_timestamps
```

Comment:
Might want to call this `timestamps_filtered`? There are two inputs, timestamps and vectors.